src/box/lua/config/utils/schema.lua

```diff
@@ -775,26 +783,52 @@ local function set_impl(schema, data, rhs, ctx)
     if is_scalar(schema) then
         walkthrough_error(ctx, 'Attempt to index a scalar value of type %s ' ..
             'by field %q', schema.type, requested_field)
-    elseif schema.type == 'record' then
+    elseif schema.type == 'record' or schema.type == 'map' then
+        -- Traverse down a record/a map that has nil/box.NULL
+        -- value to create intermediate tables on the path to
+        -- `rhs` and to validate the path against the schema.
         walkthrough_enter(ctx, requested_field)
-        local field_def = schema.fields[requested_field]
-        if field_def == nil then
-            walkthrough_error(ctx, 'No such field in the schema')
+
+        -- A schema of the requested field.
+        local field_def
+        if schema.type == 'record' then
+            field_def = schema.fields[requested_field]
+            if field_def == nil then
+                walkthrough_error(ctx, 'No such field in the schema')
+            end
+        else
+            assert(schema.type == 'map')
+            field_def = schema.value
         end
+        assert(field_def ~= nil)
 
-        walkthrough_assert_table(ctx, schema, data)
-        local field_value = data[requested_field] or {}
-        table.remove(ctx.journey, 1)
-        data[requested_field] = set_impl(field_def, field_value, rhs, ctx)
-        return data
-    elseif schema.type == 'map' then
-        walkthrough_enter(ctx, requested_field)
-        local field_def = schema.value
+        -- Current value of the requested field.
+        local field_value
+        if data ~= nil then
+            walkthrough_assert_table(ctx, schema, data)
+            field_value = data[requested_field]
+        end
 
-        walkthrough_assert_table(ctx, schema, data)
-        local field_value = data[requested_field] or {}
+        -- The value of the field may be changed to `rhs` or
+        -- from nil/box.NULL to a newly created table.
         table.remove(ctx.journey, 1)
-        data[requested_field] = set_impl(field_def, field_value, rhs, ctx)
+        local new_field_value = set_impl(field_def, field_value, rhs, ctx)
+
+        if type(new_field_value) == 'nil' then
+            -- Delete the field. Keeps nil and box.NULL, modifies
+            -- a table.
+            if data ~= nil then -- table
+                data[requested_field] = nil
+            end
+        else
```

Comment:
It would be nice to drop a comment, mentioning that `nil` branch is here to avoid creation of an empty table.
Otherwise, it is unclear at a first glance why the entire chunk was not simplified down to something like:
```lua
data = data or {}
data[requested_field] = new_field_value
```